### PR TITLE
FEAT: Complex data type support- uniqueidentifier

### PR DESCRIPTION
**Pull request by gargsaumya**
### Work Item / Issue Reference  

> [AB#34945](https://sqlclientdrivers.visualstudio.com/c6d89619-62de-46a0-8b46-70b92a84d85e/_workitems/edit/34945)

> GitHub Issue: #241 

-------------------------------------------------------------------
### Summary   
  

This pull request adds comprehensive support for Python's `uuid.UUID` type when interacting with SQL Server `UNIQUEIDENTIFIER` columns. It updates both the parameter binding and result retrieval logic to handle UUIDs as native Python objects, ensuring correct type mapping, byte order, and conversion in both directions. Additionally, new tests are introduced to verify correct insertion and retrieval of UUID values, including edge cases with `None`.

**UUID/UNIQUEIDENTIFIER Support Improvements:**

* Added logic in `mssql_python/cursor.py` to map Python `uuid.UUID` objects to the correct SQL types and convert them to raw bytes before sending to SQL Server. [[1]](diffhunk://#diff-deceea46ae01082ce8400e14fa02f4b7585afb7b5ed9885338b66494f5f38280R429-R437) [[2]](diffhunk://#diff-deceea46ae01082ce8400e14fa02f4b7585afb7b5ed9885338b66494f5f38280R777-R778)
* Updated the C++ binding layer (`mssql_python/pybind/ddbc_bindings.cpp`) to correctly marshal UUIDs between Python and SQL Server, including proper byte order conversion and error handling for incorrect UUID sizes.
* Enhanced result fetching in the C++ layer to convert SQL Server `UNIQUEIDENTIFIER` columns back into Python `uuid.UUID` objects, both for single-row and batch fetches, and to handle `NULL` values gracefully. [[1]](diffhunk://#diff-dde2297345718ec449a14e7dff91b7bb2342b008ecc071f562233646d71144a1L2202-R2249) [[2]](diffhunk://#diff-dde2297345718ec449a14e7dff91b7bb2342b008ecc071f562233646d71144a1L2593-R2639)

**Testing Enhancements:**

* Added new tests in `tests/test_004_cursor.py` to verify correct insertion, retrieval, and handling of `None` values in UUID columns, as well as round-trip integrity for multiple UUIDs. [[1]](diffhunk://#diff-82594712308ff34afa8b067af67db231e9a1372ef474da3db121e14e4d418f69R6827-R6923) [[2]](diffhunk://#diff-82594712308ff34afa8b067af67db231e9a1372ef474da3db121e14e4d418f69R17)

### CODE REVIEW

File: mssql_python/pybind/ddbc_bindings.cpp [hunk 1]
```diff
@@ -504,7 +504,32 @@ SQLRETURN BindParameters(SQLHANDLE hStmt, const py::list& params,
                 break;
             }
             case SQL_C_GUID: {
-                // TODO
+                py::object uuid_obj = param;
+                py::object uuid_cls = py::module_::import("uuid").attr("UUID");
+                if (!py::isinstance(uuid_obj, uuid_cls)) {
+                    ThrowStdException(MakeParamMismatchErrorStr(paramInfo.paramCType, paramIndex));
+                }
+                py::bytes uuid_bytes = uuid_obj.attr("bytes").cast<py::bytes>();
```
Comment (sumitmsft) [inline mssql_python/pybind/ddbc_bindings.cpp]: UUIDs are manually packed and unpacked between Python and ODBC’s SQLGUID structure using bitwise operations and direct memory copying. This approach is very sensitive to byte order (endianness), and if handled incorrectly, can lead to corrupted UUIDs and mismatches between what’s written and read from the database (because you need to match the exact byte order expected by both Python and ODBC—otherwise, your UUIDs will be incorrect.)

What I want you to comment and validate - that packing from Python’s `uuid.UUID` to ODBC’s `SQLGUID` always uses the correct byte order as required by the ODBC specification (little-endian for Data1, Data2, Data3; direct copy for Data4).

File: mssql_python/pybind/ddbc_bindings.cpp [hunk 2]
```diff
@@ -2957,9 +2988,15 @@ SQLRETURN FetchBatchData(SQLHSTMT hStmt, ColumnBuffers& buffers, py::list& colum
                     break;
                 }
                 case SQL_GUID: {
-                    row.append(
-                        py::bytes(reinterpret_cast<const char*>(&buffers.guidBuffers[col - 1][i]),
-                                  sizeof(SQLGUID)));
+                    SQLGUID* guidValue = &buffers.guidBuffers[col - 1][i];
+                    std::vector<char> guid_bytes(16);
+                    std::memcpy(guid_bytes.data(), guidValue, sizeof(SQLGUID));
```
Comment (sumitmsft) [inline mssql_python/pybind/ddbc_bindings.cpp]: OK, so based on my understanding, we are copying the raw memory of the SQLGUID struct directly into a bytes array. 
Python’s uuid.UUID(bytes=...) expects the 16 bytes in big-endian order.
But, ODBC’s SQLGUID stores some fields (Data1, Data2, Data3) in little-endian order.

I guess the direct memcpy will only work if the memory layout matches Python’s expectation, which is NOT safe or portable. This means that one some platforms, the resulting UUID in Python could have the wrong byte order and not match the value in the database. Do you agree? 

If we reconstruct the UUID bytes explicitly, converting the little-endian fields to big-endian order before passing to Python would make sure we always have a right data with us. 

I guess you've done similar stuff in SQLGetData_wrap()

Please correct me if my understanding is different from your logic.

File: mssql_python/pybind/ddbc_bindings.cpp [hunk 3]
```diff
@@ -504,7 +504,32 @@ SQLRETURN BindParameters(SQLHANDLE hStmt, const py::list& params,
...
             case SQL_C_GUID: {
-                // TODO
+                py::object uuid_obj = param;
+                py::object uuid_cls = py::module_::import("uuid").attr("UUID");
+                if (!py::isinstance(uuid_obj, uuid_cls)) {
+                    ThrowStdException(MakeParamMismatchErrorStr(paramInfo.paramCType, paramIndex));
+                }
+                py::bytes uuid_bytes = uuid_obj.attr("bytes").cast<py::bytes>();
+                if (PyBytes_GET_SIZE(uuid_bytes.ptr()) != 16) {
+                    ThrowStdException("UUID binary data must be exactly 16 bytes long.");
```
Comment (sumitmsft) [inline mssql_python/pybind/ddbc_bindings.cpp:514]: If the UUID is malformed (not 16 bytes), an exception is thrown but not logged, and details are not included in any log output. This may hinder debugging.
Can we Log the actual parameter and index before throwing, and ensure error messages do not leak sensitive content.

File: mssql_python/pybind/ddbc_bindings.cpp [hunk 4]
```diff
@@ -504,7 +504,32 @@ SQLRETURN BindParameters(SQLHANDLE hStmt, const py::list& params,
...
+                }
+                const unsigned char* b = reinterpret_cast<const unsigned char*>(PyBytes_AS_STRING(uuid_bytes.ptr()));
+                SQLGUID* guid_data_ptr = AllocateParamBuffer<SQLGUID>(paramBuffers);
```
Comment (sumitmsft) [inline mssql_python/pybind/ddbc_bindings.cpp:516]: `AllocateParamBuffer<SQLGUID>` and `AllocateParamBuffer<SQLLEN>` (**below**) allocate memory for each UUID. Ensure these are released appropriately after use, especially in error paths. If not, this can lead to leaks.

File: tests/test_004_cursor.py [hunk 5]
```diff
@@ -6942,6 +6943,103 @@ def test_money_smallmoney_invalid_values(cursor, db_connection):
         drop_table_if_exists(cursor, "dbo.money_test")
         db_connection.commit()
 
+def test_uuid_insert_and_select_none(cursor, db_connection):
```
Comment (sumitmsft) [inline tests/test_004_cursor.py:6947]: Check for the following tests:
invalid UUID values (bad length, not a UUID instance, wrong type).
duplicate UUIDs (primary key violation).
extreme valid UUIDs (e.g., all zeros, all ones).

**Review by sumitmsft — CHANGES_REQUESTED** [review]
Left a few comments

File: mssql_python/pybind/ddbc_bindings.cpp (hunk 4 above)
Comment (gargsaumya) [reply mssql_python/pybind/ddbc_bindings.cpp:516]: AllocateParamBuffer uses scope-managed buffers; all allocated memory is cleaned up appropriately when the parameter buffers go out of scope, so no leaks are expected.

File: mssql_python/pybind/ddbc_bindings.cpp (hunk 2 above)
Comment (gargsaumya) [reply mssql_python/pybind/ddbc_bindings.cpp]: Right.
Our current solution explicitly reconstructs the UUID bytes from SQLGUID, converting Data1, Data2, Data3 to big-endian order before passing to Python. This guarantees that the Python uuid.UUID object always matches the value stored in the database, regardless of platform or compiler endianness.

File: tests/test_004_cursor.py (hunk 5 above)
Comment (gargsaumya) [reply tests/test_004_cursor.py:6947]: Added tests for:

1. Invalid UUID values (wrong type, bad length, non-UUID objects)
2. Duplicate UUIDs (primary key violation)
3. Extreme valid UUIDs (all zeros, all ones)

All tests pass, and the driver behaves correctly in each case.

**Review by bewithgaurav — APPROVED** [review]
